### PR TITLE
feat(messaging): Make SignalManager re-entrant

### PR DESCRIPTION
**Pull request by mykola-vrmchk**

### CODE REVIEW

**Comment by mykola-vrmchk** [comment]
It should be tested more

File: packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts [hunk 1]
```diff
@@ -59,26 +64,89 @@ export class WebsocketSignalManager implements SignalManager {
...
+      [...this._servers.values()].map((server: SignalClient) =>
+        server.sendMessage({ author, recipient, payload }).catch((err) => log(err))
+      )
+    );
+  }
+
+  async subscribeMessages(peerId: PublicKey): Promise<void> {
+    log(`Subscribed for message stream peerId=${peerId}`);
+    assert(!this._closed, 'Closed');
+    this._subscribedMessages.add(peerId);
```
Comment (mykola-vrmchk) [inline packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts:141]: @dmaretskyi what do you think about it?

File: packages/core/mesh/messaging/src/messenger.ts [hunk 2]
```diff
@@ -23,6 +23,9 @@ export interface MessengerOptions {
   timeout?: number;
 }
 
+/**
+ * Reliable messenger that works trough Signal Servers network.
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/messenger.ts:27]:
```suggestion
 * Reliable messenger that works through the signal network.
```

File: packages/core/mesh/messaging/src/messenger.ts [hunk 3]
```diff
@@ -37,22 +40,43 @@ export class Messenger {
 
   private readonly _receivedMessages = new ComplexSet<PublicKey>((key) => key.toHex());
 
-  private readonly _subscriptions = new EventSubscriptions(); // TODO(burdon): Not released.
+  private readonly _subscriptions = new EventSubscriptions();
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/messenger.ts:43]: change to Context please

File: packages/core/mesh/messaging/src/messenger.ts [hunk 4]
```diff
@@ -37,22 +40,43 @@ export class Messenger {
...
+
+    this.open();
+  }
+
+  open() {
+    if (!this._closed) {
+      return;
+    }
+    this._subscriptions.add(
+      this._signalManager.onMessage.on(async (message) => {
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/messenger.ts:61]:
```suggestion
      this._signalManager.onMessage.on(this._ctx, async (message) => {
```

File: packages/core/mesh/messaging/src/messenger.ts (hunk 3 above)
Comment (dmaretskyi) [reply packages/core/mesh/messaging/src/messenger.ts:43]: you might have to re-create the context to make it reentrant

File: packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts [hunk 5]
```diff
@@ -59,26 +64,89 @@ export class WebsocketSignalManager implements SignalManager {
...
+
+    await Promise.all(
+      [...this._servers.values()].map((signalClient: SignalClient) => signalClient.subscribeMessages(peerId))
+    );
+  }
+
   private _scheduleReconcile() {
-    if (this._destroyed) {
+    if (this._closed) {
```
Comment (dmaretskyi) [inline packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts:149]: You should be able to replace this scheduling logic with DeferredTask

Comment (dmaretskyi) [reply packages/core/mesh/messaging/src/signal-manager/websocket-signal-manager.ts:149]: see example in replication-extension

**Review by dmaretskyi — APPROVED** [review]

**Pull request merged**